ethcontract/src/batch.rs

```diff
@@ -0,0 +1,89 @@
+//! Module containing components to batch multiple contract calls
+//! into a single request to the Node.
+
+use futures::channel::oneshot::{channel, Sender};
+use web3::{
+    error::Error as Web3Error,
+    helpers::{self},
+    types::{BlockId, BlockNumber, Bytes, CallRequest},
+    BatchTransport as Web3BatchTransport,
+};
+
+/// Struct allowing to batch multiple calls into a single Node request
+pub struct CallBatch<T: Web3BatchTransport> {
+    inner: T,
+    requests: Vec<(Request, CompletionHandler)>,
+}
+
+type Request = (CallRequest, Option<BlockId>);
+type CompletionHandler = Sender<Result<Bytes, Web3Error>>;
+
+impl<T: Web3BatchTransport> CallBatch<T> {
+    /// Create a new instance from a BatchTransport
+    pub fn new(inner: T) -> Self {
+        Self {
+            inner,
+            requests: Default::default(),
+        }
+    }
+
+    /// Adds a call request to the current batch. The resulting future can only resolve after
+    /// the batch has been resolved via `execute_all`.
+    /// Panics, if the batch is dropped before executing.
+    pub fn push(
+        &mut self,
+        call: CallRequest,
+        block: Option<BlockId>,
+    ) -> impl std::future::Future<Output = Result<Bytes, Web3Error>> {
```

Comment:
Is there a reason this was implemented as `fn x(...) -> impl Future<Output = O>` instead of `async fn x(...) -> O`? I would suspect there wouldn't really be any difference.

**edit**: Now that I think of it, it makes a clear separation between the future init code and the actual future which is important for not causing races (like trying to add a request after the batch is exectuting because the future was never polled once.).

Reply:
As you said, the reason is that we need some code to run before returning the future which can we cannot do in the future itself (which would be created by an async block) because futures do nothing unless polled.
Since this is uncommon we might want to add a comment to the method explaining that.

Reply:
My motivation was that this way the Future doesn't have the same lifetime as `&mut self` (which is the case in the default async impl) meaning that I would only to use the batch once before the future is awaited (basically for the reason both of you were referring to). WIll add a comment...